File: dataAnalysis/wordsScience/one-edit.go
package main

import (
	"strings"
	"fmt"
)

const Alphabet = "abcdefghijklmnopqrstuvwxyz"

func Map(vs []string, f func(string) string) []string {
	vsm := make([]string, len(vs))
	for i, v := range vs {
		vsm[i] = f(v)
	}
	return vsm
}

func head(vs []string) string {
	return vs[0]
}

func group(vs []string) map[string]int {
	vsg := make(map[string]int)
	for _, v := range vs {
		_, had := vsg[v]
		if had {
			vsg[v]++
		} else {
			vsg[v] = 1
		}
	}
	return vsg
}

func keys(m map[string]int) []string {
	keys := make([]string, len(m))
	i := 0
	for k := range m {
		keys[i] = k
		i++
	}
	return keys
}

func unique(s []string) []string {
	return keys(group(s))
}

func deletes(s string) []string  {
	ds := make([]string, len(s))
	for i := range s {
		ds[i] = s[0:i] + s[i+1:]
	}
	return ds
}

func transposes(s string) []string {
	ts := make([]string, len(s) - 1)
	for i:=0; i<len(s) - 1; i++ {
		c := []byte(s)
		c[i], c[i+1] = c[i+1], c[i]
		ts[i] = string(c)
	}
	return ts
}

func replaces(s string) []string  {
	alphabetLen := len(Alphabet)
	length := len(s) * alphabetLen
	rs := make([]string, length)
	for i := range s {
		for j := range Alphabet {
			c := []byte(s)
			c[i] = Alphabet[j]
			rs[i+j] = string(c)
		}
	}
	return rs
}

func inserts (s string) []string {
	alphabetLen := len(Alphabet)
	length := (len(s) + 1) * alphabetLen
	is := make([]string, length)
	alphabet := []byte(Alphabet)
	for i := range s {
		for j := range alphabet {
			is[i+j] = string(s[:i] + string(alphabet[j]) + s[i:])
		}
	}
	return is
}


func edits1(word string) []string {
	lower := strings.ToLower(word)
	ds := deletes(lower)
	fmt.Println(len(ds))
	ts := transposes(lower)
	fmt.Println(len(ts))
	rs := replaces(lower)
	fmt.Println(len(rs))
	is := inserts(lower)
	fmt.Println(len(is))
	var words []string
	words = append(words, ds...)
	words = append(words, ts...)
	words = append(words, rs...)
	words = append(words, is...)
	return unique(words)
}

func main() {
	fmt.Println(edits1("hi"))
}

File: dataAnalysis/wordsScience/one-edit.hs
import Data.Char (toLower)
import Data.List (group, sort)

edits1 :: String -> [String]
edits1 word = unique $ deletes ++ transposes ++ replaces ++ inserts
  -- splits 从 0 开始 take 和 drop 构造 "hi" -> [("", "hi"), ("h", "i"), ("hi", "")]
  -- "hello" -> [("", "hello"), ("h", "ello"), ("he", "llo"), ("hel", "lo"), ("hell", "o"), ("hello", "")]
  where splits = [ (take i word', drop i word') | i <- [0..length word'] ]
        -- 删掉其中一个字母
        deletes = [ a ++ (tail b) | (a, b) <- splits, (not.null) b ]
        -- 前后互换顺序 只更新 b 中前两个字母的顺序，然后拼接即可
        transposes = [ a ++ [b!!1] ++ [head b] ++ (drop 2 b) | (a,b) <- splits, length b > 1 ]
        -- 替换一个字母 只从 b 中替换，然后拼接即可
        replaces = [ a ++ [c] ++ (drop 1 b) | (a, b) <- splits, c <- alphabet, (not.null) b ]
        -- 插入一个字母
        inserts = [ a ++ [c] ++ b | (a, b) <- splits, c <- alphabet ]
        alphabet = ['a'..'z']
        word' = map toLower word

unique :: [String] -> [String]
-- (map head).group.sort $ []
unique = map head.group.sort

main = print $ edits1 "hi"

edits1' :: [String] -> [String]
edits1' ls = unique $ concat $ map edits1 ls

editsN :: String -> Int -> [String]
editsN word n = iterate edits1' (edits1 word) !! n
